### DISCUSSION TITLE
Querying related tables with limit

### DISCUSSION THREAD

**Discussion by HasanAlqaisi** [simolus3/drift#2561] [Q&A] [answered]
Hi,

I have three tables:
1. PatientsTable
2. AllergiesTable
3. TagsTable

**PatientsTable** has n:m relationship with both **AllergiesTable** & **TagsTable**, so there are 2 other intermediary tables.

I am finding difficulty on how to write a dart query that fetch all patients and their allergies & tags with **LIMIT**

I found out that discussion [#2253](https://github.com/simolus3/drift/discussions/2253) , it works well without LIMIT which is not my case.

I don't want to use watch like explained in the doc example that's why I tried the above strategy in the discussion.

Here's the code:

```dart
  Future<List<PatientLocal>> getPatients(GetPatientsParams params) async {
    final selectQuery = select(patientsTable);

    if (params.searchName != null) {
      selectQuery.where((tbl) => tbl.name.contains(params.searchName!));
    }

    selectQuery.orderBy([
      (tbl) => OrderingTerm.desc(tbl.identifier),
      (tbl) => OrderingTerm.desc(tbl.serverId, nulls: NullsOrder.first),
    ]);

    final joinQuery = selectQuery.join([
      leftOuterJoin(
        patientAllergyTable,
        patientAllergyTable.localPatientId.equalsExp(patientsTable.localId),
      ),
      leftOuterJoin(allergiesTable,
          allergiesTable.localId.equalsExp(patientAllergyTable.localAllergyId)),
      leftOuterJoin(
        patientTagTable,
        patientTagTable.localPatientId.equalsExp(patientsTable.localId),
      ),
      leftOuterJoin(
          tagsTable, tagsTable.localId.equalsExp(patientTagTable.localTagId)),
    ]);

    joinQuery.limit(params.pageSize,
        offset: (params.page - 1) * params.pageSize);

    final idsToPatients = <int, PatientLocal>{};

    await joinQuery.map((row) {
      final patientLocal = row.readTable(patientsTable);

      idsToPatients.putIfAbsent(patientLocal.localId, () => patientLocal);

      // ...more operations related to reading allergy & tag 
    }).get();

    return idsToPatients.values.toList();
  }
```
If the limit was 12, The above code is not giving me 12 patients.
I think this is because in the **map**  method, patients can be duplicated depending on the relations.

So my question is; Is there any other ways to get related data without using watch() using drift dart API ?

**Comment by simolus3** [comment, marked as answer]
To which table do you want the limit to apply to? E.g. do you want 12 patients, but load every allergy and tag for every one of those 12 patients?

In that case, I think an `IS IN` subquery to load patient ids might be an easy solution that doesn't require you to rewrite the query much. 

```dart
  Future<List<PatientLocal>> getPatients(GetPatientsParams params) async {
    final patientIds = selectOnly(patientsTable)
      ..addColumns([patientsTable.localId])
      ..orderBy([
        OrderingTerm.desc(patientsTable.identifier),
        OrderingTerm.desc(patientsTable.serverId, nulls: NullsOrder.first),
      ])
      ..limit(params.pageSize, offset: params.pageSize);

    if (params.searchName != null) {
      patientIds.where(patientsTable.name.contains(params.searchName!));
    }

    // Now use patientIds query to only filter 12 patients to begin with.
    final selectQuery = select(patientsTable)
      ..where((row) => row.localId.isInQuery(patientIds));
    selectQuery.orderBy([
      (tbl) => OrderingTerm.desc(tbl.identifier),
      (tbl) => OrderingTerm.desc(tbl.serverId, nulls: NullsOrder.first),
    ]);

    // rest of join can stay the same, except that you don't need a limit anymore
```

The idea is that we first query which patients to load in another query. This query is them embedded into the main one via [`isInQuery`](https://drift.simonbinder.eu/api/drift/expression/isinquery) to only select the 12 patients of the current page.

**Reply by HasanAlqaisi** [reply]
This is exactly what I wanted! Thank you so much.
